### PR TITLE
feat(gar): prune gateways only when necessary PE-7147

### PR DESCRIPTION
**Pull request by arielmelendez**

### CODE REVIEW

File: src/gar.lua [hunk 1]
```diff
@@ -838,6 +847,11 @@ function gar.updateGatewayStats(address, gateway, stats)
 	assert(stats.failedConsecutiveEpochs, "failedConsecutiveEpochs is required")
 	assert(stats.passedConsecutiveEpochs, "passedConsecutiveEpochs is required")
 
+	if stats.failedConsecutiveEpochs >= gar.getSettings().operators.failedEpochCountMax then
+		-- Ensure pruning takes place at the next opportunity
+		NextGatewaysPruneTimestamp = 0
+	end
```
Comment (dtfiedler) [inline src/gar.lua:853]: this happens on `distribute` which does not run automatically like `prune`. will not hurt, but the gateway won't be moved to `leaving` until the next `Tick` comes in

File: src/gar.lua [hunk 2]
```diff
@@ -96,6 +96,9 @@ GatewayRegistrySettings = GatewayRegistrySettings
 		},
 	}
 
+--- @type Timestamp|nil
+NextGatewaysPruneTimestamp = NextGatewaysPruneTimestamp or 0
```
Comment (fedellen) [inline src/gar.lua:100]: whats the main difference between 0 and nil here for the logic? 

```suggestion
--- @type Timestamp
NextGatewaysPruneTimestamp = NextGatewaysPruneTimestamp or 0
```

File: src/gar.lua [hunk 3]
```diff
@@ -895,12 +907,23 @@ function gar.pruneGateways(currentTimestamp, msgId)
 		delegateStakeWithdrawing = 0,
 		stakeSlashed = 0,
 	}
+	if not NextGatewaysPruneTimestamp or currentTimestamp < NextGatewaysPruneTimestamp then
+		-- No known pruning work to do
+		return result
+	end
```
Comment (fedellen) [inline src/gar.lua:913]: ah, 0 must be truthy.

**Review by fedellen — APPROVED** [review]

File: src/gar.lua (hunk 1 above)
Comment (arielmelendez) [reply src/gar.lua:853]: Good to know. Will remove to avoid complexity.

File: src/gar.lua (hunk 2 above)
Comment (arielmelendez) [reply src/gar.lua:100]: It's an initial condition in this case. When we deploy updates to the source, globals will have previous values available to assign here. 0 indicates "definitely prune next time pruning happens". `nil` indicates that there's nothing to prune. 0 is more conservative when we're not sure what the deal should be.

File: src/gar.lua (hunk 3 above)
Comment (arielmelendez) [reply src/gar.lua:913]: Correct, 0 is truthy in Lua

File: src/gar.lua (hunk 1 above)
Comment (arielmelendez) [reply src/gar.lua]: 629ba1f

**Pull request merged**